### ISSUE TITLE
Are we really use 1-indexed data to train model (looks like not, from dataloader)? 

### ISSUE THREAD

**Issue by Cli98** [zylo117/Yet-Another-EfficientDet-Pytorch#249]
Hi @zylo117 ,

I'm currently trying to make some post analysis for my result. As I mentioned before, the performance does not meet my expectation.

During the time I conduct my analysis, I found that after fetching my annotation from dataloader, my label becomes 0-indexed, instead of 1-indexed. This is clearly not what should be expected as you mentioned. 

<img width="648" alt="image" src="https://user-images.githubusercontent.com/19630616/80932550-f6476280-8d8d-11ea-92d1-794277b6ccae.png">

I carefully read your data loader implementation and found in dataset.py, L89-L90 you defined a function "coco_label_to_label", and then you call this function in order to retrieve category id for each bbox. But when you initialize this dictionary, you assigned with len(self.classes) (at Line 33) and thus index becomes zero based. 

Check following plot. I print out   self.classes, self.coco_labels, self.coco_labels_inverse. The issue is with the last one. For given c['id'], index changes to c['id']-1, thus zero based.

<img width="907" alt="image" src="https://user-images.githubusercontent.com/19630616/80934326-b2f0f200-8d95-11ea-9709-9b5c66e01aa8.png">

Since this is not what you described in your Readme, I would like to know if this is the expected behavior or a bug.

Thank you,

**Comment by zylo117** [comment]
I see, it's a bug succeed from the original label convertion.
It does affect training coco from scratch. But it shouldn't have affect training on a custom dataset.

**Comment by zylo117** [comment]
it's fixed. https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/063b845783a9e7fe5c30bc86dbf5f9a52305e42b

**Comment by Cli98** [comment]
> it's fixed. [063b845](https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/063b845783a9e7fe5c30bc86dbf5f9a52305e42b)

@zylo117 Great. And can you tell how will it affect result? Do you recommend me to retrain my model?

**Comment by zylo117** [comment]
long story. mainly because coco category ida are sparse.
if you are training coco, yes, you need to retrain.
but if you are training on custom dataset, no.

**Comment by Cli98** [comment]
> long story. mainly because coco category ida are sparse.
> if you are training coco, yes, you need to retrain.
> but if you are training on custom dataset, no.

Thanks.

**Comment by zylo117** [comment]
@Cli98 sorry, it was not fixed. now it is.  https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/0b16c7c2194970ccdee843099322024bff8a9636

**Comment by Cli98** [comment]
@zylo117 I do not think this fix is valid as labels range in [0,89] now. You can try to print out a['category_id'] VS a['category_id']-1
<img width="200" alt="image" src="https://user-images.githubusercontent.com/19630616/80977450-bc5c7780-8df2-11ea-9933-dba5cb9db048.png">

As far as I can tell, a['category_id'] is 1-index based. Please let me know the reason why you choose a['category_id']-1

**Comment by Cli98** [comment]
> @Cli98 sorry, it was not fixed. now it is. [0b16c7c](https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/0b16c7c2194970ccdee843099322024bff8a9636)

@zylo117 And I'm also confusing when should be 0-indexed or 1-indexed. You previously explained that it is due to the stage. Any reference for this?

**Comment by zylo117** [comment]
the annotation file is one-indexed. but the program is zero-indexed.

**Comment by Cli98** [comment]
> the annotation file is one-indexed. but the program is zero-indexed.

This does not make sense to me. What's the meaning of "program is zero-indexed"? Shouldn't annotations and programs both share same indexing? 

If you set program to "zero-indexed", what's the point to care whether annotation is 0-indexed or 1-indexed? You do not even need to pull out the code and change it. And your current implementation is the same as before.

**Comment by zylo117** [comment]
because coco-style annotation is one-indexed. It'd be better to keep it unchanged to make labelling easier.
The training program is always zero-indexed. But there are mismatch bug, which is not about one-indexed, so I need to fix it.
you should combine these two commits. 
https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/063b845783a9e7fe5c30bc86dbf5f9a52305e42b
https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/0b16c7c2194970ccdee843099322024bff8a9636

**Comment by Cli98** [comment]
> because coco-style annotation is one-indexed. It'd be better to keep it unchanged to make labelling easier.
> The training program is always zero-indexed. But there is a mismatch bug, which is not about one-indexed, so I need to fix it.
> you should combine these two commits.
> [063b845](https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/063b845783a9e7fe5c30bc86dbf5f9a52305e42b)
> [0b16c7c](https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/0b16c7c2194970ccdee843099322024bff8a9636)

This does not solve my concern. According to your modification 0b17c7c, In training you set labeling to [0,89] then your prediction will range in [0,89] in test phase.  What will happen when you evaluate with ground truth with label ranges in [1,90]? There is clearly a mismatch anyway and coco_eval cannot take care of it. How you plan to solve it? I did not see any code take care of this.

**Comment by zylo117** [comment]
Nooooooooo.
My original intention is to fix coco label mapping mismatch issue, which has nothing to do with zero-index/ one-index. But I fixed it with introducing another bug that I forget to set it back to zero-index.

**Comment by Cli98** [comment]
> Nooooooooo.
> My original intention is to fix coco label mapping mismatch issue, which has nothing to do with zero-index/ one-index. But I fixed it with introducing another bug that I forget to set it back to zero-index.

We are not on the same page... Let me take one step further, in prediction we target on car(let's say its label is 3 in coco but 2 in training program). So after you predict, if correctly, the output is 2. When you load annotation from coco, you find the corresponding annotation is 3. My question is, Is it a bug or not?
